src/Build/BackEnd/Node/RarNode.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Server;
+
+namespace Microsoft.Build.Execution
+{
+    public sealed class RarNode : INode
+    {
+        /// <summary>
+        /// The amount of time to wait for the client to connect to the host.
+        /// </summary>
+        private const int ClientConnectTimeout = 60000;
+
+        public NodeEngineShutdownReason Run(bool nodeReuse, bool lowPriority, out Exception shutdownException)
+        {
+            shutdownException = null;
+            using CancellationTokenSource cts = new CancellationTokenSource();
+            string pipeName = CommunicationsUtilities.GetRarPipeName(nodeReuse, lowPriority);
+            RarController controller = new RarController(pipeName, NamedPipeUtil.CreateNamedPipeServer);
+
+            Task<int> rarTask = controller.StartAsync(cts.Token);
+
+            Handshake handshake = NodeProviderOutOfProc.GetHandshake(nodeReuse, enableLowPriority: lowPriority, specialNode: true);
+            Task<NodeEngineShutdownReason> msBuildShutdown = RunShutdownCheckAsync(handshake, cts.Token);
+
+            int index = Task.WaitAny(msBuildShutdown, rarTask);
```

Comment:
I don't think so. 
`msBuildShutdown` - this can
`rarTask` - has 15 minutes timeout between connections, so I don't think so.

And if we rely on fact that the rarTask will not hang, then we can be sure that this will be ok.

Reply:
I think you're right, but I'd feel a little better if you included an explicit timeout here.

Reply:
I agree with @Forgind 's recommendation of specifying timeouts or passing cancellation token handles

Reply:
So there will be timeout, which will shutdown the tasks (and the node). This timeout will happen after some specific time, without regard what is currently RAR node doing? Is this what you have in mind? Just asking for clarification.

Reply:
Added timeout to 1 hour.